### PR TITLE
[Inductor] Add FlyDSL gemm template integration

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Add an experimental FlyDSL gemm template backend for ROCm Inductor autotuning, including async compilation, scheduling, runtime gating, and focused tests.

### CODE REVIEW

File: torch/_inductor/codegen/flydsl/flydsl_utils.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import subprocess
+from pathlib import Path
+
+from torch.backends import cuda as _cuda
+
+
+log = logging.getLogger(__name__)
+
+
+def _flydsl_runtime_unavailable_reason() -> str | None:
```
Comment (jataylo) [inline torch/_inductor/codegen/flydsl/flydsl_utils.py:15]: Reimplementation from torch/_native/flydsl_utils.py do we need both? can we share some utils logic across native and inductor?

File: torch/_inductor/codegen/flydsl/flydsl_utils.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        Path(next(iter(flydsl_spec.submodule_search_locations)))
+        / "_mlir"
+        / "_mlir_libs"
+        / "libfly_jit_runtime.so"
+    )
+    if not runtime_so.exists():
+        return f"missing FlyDSL runtime shared library `{runtime_so}`"
+
+    try:
+        ldd = subprocess.run(
```
Comment (jataylo) [inline torch/_inductor/codegen/flydsl/flydsl_utils.py]: Ldd is linux specific, if windows support is an eventuality need to rethink this. We could guard on platform.system() != "Linux" which may be cleaner

File: torch/_inductor/codegen/flydsl/flydsl_kernel.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            )
+        for arg_def, call_arg, arg_type in zip(
+            orig_arg_defs, orig_call_args, orig_arg_types
+        ):
+            if arg_def.full_name() in self._seen_input_args:
+                continue
+            call_args.append(call_arg)
+            arg_types.append(arg_type)
+
+        wrapper.generate_kernel_call(name, call_args, triton=True, arg_types=arg_types)
```
Comment (jataylo) [inline torch/_inductor/codegen/flydsl/flydsl_kernel.py]: IIUC `triton=True` is required to emit the wrapper code properly, but flag is confusing as we are not actually using triton here. We could potentially refactor this.

File: test/inductor/test_flydsl_template.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        self.assertIn("import torch", imports)
+        self.assertIn("import flydsl.compiler as flyc", imports)
+        self.assertIn("import flydsl.expr as fx", imports)
+
+    @unittest.skipUnless(HAS_FLYDSL, "requires flydsl")
+    @unittest.skipUnless(torch.cuda.is_available(), "CUDA/ROCm not available")
+    @unittest.skipIf(torch.version.hip is None, "requires ROCm")
+    @torch._inductor.config.patch(
+        max_autotune_gemm=True,
+        max_autotune_gemm_backends="FLYDSL",
```
Comment (jataylo) [inline test/inductor/test_flydsl_template.py:328]: Perhaps we could also enable some of the regular inductor UTs in test_max_autotune to parameterise with FlyDSL backend to run more UTs.

**Review by jataylo — APPROVED** [review]
Some comments but nothing blocking, LGTM

File: torch/_inductor/codegen/flydsl/flydsl_utils.py (hunk 2 above)
Comment (jgong5) [inline torch/_inductor/codegen/flydsl/flydsl_utils.py]: Also, the checking logic is inconsistent with the eager mode. Are we missing something or it is on purpose, and why?

File: torch/_inductor/kernel/templates/flydsl_mm.py.jinja [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            if old_value is None:
+                os.environ.pop(key, None)
+            else:
+                os.environ[key] = old_value
+
+
+def _hgemm_splitk_mm(output, mat1, mat2, stream):
+    k = mat1.shape[-1]
+    if mat2.shape[-1] == k:
+        mat2_nk = mat2
```
Comment (jgong5) [inline torch/_inductor/kernel/templates/flydsl_mm.py.jinja]: **[by claude]** Likely wrong result when `N == K`. The `mm.py` gate only admits the transposed-RHS case (`mat2_stride[0] == 1`), so `mat2` here is always the `[K, N]` view of the original `[N, K]` tensor and should *always* be transposed back to `[N, K]`. But this branch disambiguates by shape: with `mat2` shaped `[K, N]`, `mat2.shape[0] == k` is always true (the intended `elif` path). When `N == K`, the `if mat2.shape[-1] == k` branch matches first and skips the transpose, feeding `b.t()` where `b` is expected -> computes `a @ b` instead of `a @ b.T` for non-symmetric inputs. Decide by stride (or unconditionally transpose, since the gate guarantees the transpose view) rather than by shape.

File: torch/_inductor/codegen/flydsl/flydsl_scheduling.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    def get_backend_features(cls, device) -> OrderedSet[BackendFeature]:
+        return OrderedSet()
+
+    @staticmethod
+    def is_flydsl_template(node: BaseSchedulerNode) -> bool:
+        return isinstance(node, SchedulerNode) and isinstance(
+            node.node, FlyDSLTemplateBuffer
+        )
+
+    def is_flydsl_fused_template(self, node: BaseSchedulerNode) -> bool:
```
Comment (jgong5) [inline torch/_inductor/codegen/flydsl/flydsl_scheduling.py]: **[by claude]** `is_flydsl_fused_template` can never return True and has no callers. It checks `isinstance(node, FusedSchedulerNode) and self.is_flydsl_template(node)`, but `is_flydsl_template` requires `isinstance(node, SchedulerNode)`; `FusedSchedulerNode` and `SchedulerNode` are sibling classes, so the conjunction is always False. This is a faithful copy of the equally-broken-and-dead `is_cutedsl_fused_template`. Since it is new code, either delete it or implement it correctly via `node.get_template_node()` (see `NVUniversalGemmScheduling.is_nv_universal_gemm_fused_template`).

File: torch/_inductor/kernel/mm.py [hunk 7]
```diff
@@ -206,6 +212,87 @@ def has_zero_dim(size) -> bool:
...
+def get_flydsl_mm_template_kwargs(
+    layout, mat1, mat2, static_shape, is_nonzero
+) -> list[dict[str, Any]]:
+    from ..template_heuristics.flydsl_gemm import get_hgemm_configs
+
+    if not (
+        static_shape
+        and is_nonzero
+        and (inductor_config.max_autotune or inductor_config.max_autotune_gemm)
+        and "FLYDSL"
```
Comment (jgong5) [inline torch/_inductor/kernel/mm.py]: **[by claude]** Backend gating is inlined here instead of following the established `use_*_template` helper pattern. Every other GEMM backend exposes a helper in `torch/_inductor/utils.py` gated on `_use_autotune_backend("...")` (`use_cutlass_template`, `use_ck_gemm_template`, `use_nv_universal_gemm_template`, ...), called from `mm.py`. Add a `use_flydsl_template(...)` helper alongside the others for consistency and discoverability, rather than hand-parsing `max_autotune_gemm_backends.split(",")`.

File: torch/_inductor/codegen/flydsl/flydsl_template.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    def maybe_append_choice(
+        self, choices: list[Any], **kwargs: Any
+    ) -> NotImplementedError | None:
+        try:
+            choices.append(self.generate(**kwargs))
+            return None
+        except NotImplementedError as e:
+            log.debug("FlyDSL template choice generation failed: %s", e)
+            return e
+        except Exception as e:
```
Comment (jgong5) [inline torch/_inductor/codegen/flydsl/flydsl_template.py:60]: **[by claude]** `except Exception` converts any integration bug into a silently-dropped autotune choice visible only at `log.debug`. When FlyDSL is explicitly requested via `max_autotune_gemm_backends="FLYDSL"` and every choice fails, the user gets no signal. Consider logging at `warning` when the backend was explicitly requested, and/or distinguishing "template genuinely not applicable" (NotImplementedError) from unexpected errors.

File: test/inductor/test_flydsl_template.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        if not flydsl_utils.runtime_available():
+            self.skipTest("FlyDSL runtime unavailable")
+
+        def fn(a, b):
+            return torch.mm(a, b.t())
+
+        for dtype in (torch.float16, torch.bfloat16):
+            with self.subTest(dtype=dtype):
+                a = torch.randn(32, 128, device="cuda", dtype=dtype)
+                b = torch.randn(128, 128, device="cuda", dtype=dtype)
```
Comment (jgong5) [inline test/inductor/test_flydsl_template.py]: **[by claude]** This is the only numeric test and it uses `N == K == 128`, which is exactly the ambiguous case in `_hgemm_splitk_mm`s shape-based transpose disambiguation (see my comment on `flydsl_mm.py.jinja`). It therefore never exercises the non-square transpose branch the backend actually targets, and cannot catch the `N == K` mis-transpose. Add a non-square case (e.g. `b = torch.randn(256, 128, ...)` -> N=256, K=128) so both the transpose path and the `N != K` disambiguation are covered.

File: torch/_inductor/codegen/flydsl/flydsl_kernel.py (hunk 3 above)
Comment (jgong5) [reply torch/_inductor/codegen/flydsl/flydsl_kernel.py]: This follows the code from CuteDSL which also has a similar comment in the code. We can refactor them together in the future.

File: torch/_inductor/template_heuristics/flydsl_gemm.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
+from dataclasses import asdict, dataclass
+
+
+@dataclass(frozen=True)
+class FlyDSLHGemmConfig:
```
Comment (jgong5) [inline torch/_inductor/template_heuristics/flydsl_gemm.py]: **[by claude]** Consider naming this dtype-agnostically for future extension. Its fields (TILE_M/N/K, STAGES, SPLIT_K, BLOCK_*_WARPS, B_TO_LDS) are not half-specific, so `FlyDSLGemmConfig` / `get_gemm_configs` would read more accurately and let a future fp8/int8 FlyDSL kernel reuse this heuristics layer. Compare CuteDSL, which names its templates by operation (`grouped_gemm_cutedsl`, `flash_attention_cutedsl_template`) and resolves dtype generically via `TORCH_TO_CUTE_DTYPE` rather than baking it into names. The vendored `hgemm_splitk` kernel is genuinely half-only, so keep that name; this is only about the PyTorch-owned glue (config/heuristics) -- the `mm.py` gate would dispatch the dtype-family to the right config-set and kernel.

File: torch/_inductor/runtime/flydsl_cache.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+import os
+from pathlib import Path
+
+
+def _cache_dir() -> Path:
+    cache_root = os.environ.get(
+        "TORCHINDUCTOR_CACHE_DIR",
+        os.path.join(os.path.expanduser("~"), ".cache", "torch_inductor"),
+    )
```
Comment (jgong5) [inline torch/_inductor/runtime/flydsl_cache.py]: Should we reuse the code from cache_dir_utils.py?

File: torch/_inductor/ir.py [hunk 12]
```diff
@@ -6439,6 +6439,38 @@ def get_outputs(self) -> list[Buffer]:
         return self.outputs
 
 
+class FlyDSLTemplateBuffer(TemplateBuffer):
```
Comment (jgong5) [inline torch/_inductor/ir.py:6508]: **[by claude]** `FlyDSLTemplateBuffer` is byte-for-byte identical to `CuteDSLTemplateBuffer` (only the name/docstring differ), and the rest of this backend is ~90% duplicated from CuteDSL (`FlyDSLBenchmarkRequest`, `async_compile.flydsl`, `flydsl_scheduling`, the Template/Caller/kernel wrapper) -- differing only in labels, imports, and the kernel-name prefix; the one genuinely new piece is `_build_flydsl_gpu_arch`. CuteDSL and NV Universal GEMM already duplicate this same skeleton, so this is the 3rd/4th copy. Please introduce a shared base for the template buffer -- e.g. a neutral `_DSLTemplateBuffer(TemplateBuffer)` holding the shared `__init__` + `get_outputs`, with `FlyDSLTemplateBuffer` and `CuteDSLTemplateBuffer` as thin subclasses -- rather than adding another copy. Ideally extend the same shared-base approach to the wrapper/kernel/template/caller/scheduling layers (parameterized by backend label, wrapper class, env-var list, buffer/prefix) and fold the existing CuteDSL/NVGemm copies in, likely as a follow-up refactor.

File: torch/_inductor/kernel/vendored_templates/flydsl/kernels/gemm_gfx950.py
Comment (jgong5) [inline torch/_inductor/kernel/vendored_templates/flydsl/kernels/gemm_gfx950.py:1]: I would suggest not to use the name prefix "h" here since hgemm might be confused with "half" or "fp16" gemm while here we actaully mean "high-performance".

File: torch/_inductor/kernel/vendored_templates/flydsl/kernels/gemm_gfx950.py [hunk 13]
```diff
@@ -0,0 +1,501 @@
...
+    name = f"hgemm_t{param.block_m}x{param.block_n}x{param.block_k}x{param.stages}"
+    name += f"_w{param.m_waves}x{param.n_waves}"
+    name += f"_gm{param.group_m}"
+    name += f"_bias{int(param.has_bias)}"
+    name += f"_ktail{int(param.has_k_tail)}"
+    name += "_nt"
+    return name
+
+
+class BlockSwizzle:
```
Comment (jgong5) [inline torch/_inductor/kernel/vendored_templates/flydsl/kernels/gemm_gfx950.py:197]: The block swizzle as well as the "barrier" and "buffer_load" functions seem to be general enough to be abstracted from flydsl instead of re-implemented here. Can we move them to flydsl?

File: torch/_inductor/kernel/templates/flydsl_mm.py.jinja [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+
+{{def_kernel("mat1", "mat2")}}
+    output = {{get_output()}}
+    global _gemm_executor, _gemm_param
+{% if MAT2_IS_NK %}
+    # The lowering passes the original contiguous B buffer for A @ B.T.
+    mat2_nk = mat2
+{% else %}
+    mat2_nk = mat2.transpose(0, 1)
```
Comment (xytpai) [inline torch/_inductor/kernel/templates/flydsl_mm.py.jinja:52]: Is this transpose calls loops kernel?
